### PR TITLE
Move cnfapp to DPDK 23.11

### PR DESCRIPTION
**Pull request by ramperher**
- Moving to DPDK 23.11 in cnfapp image, changing the way DPDK is built, which moved from make to meson. Installing required packages for this to work and running proper commands
  - same work will be done for testpmd image in a separate change since it requires a reimplementation of Saravanan's patch
- Support for `EAL_IGB_UIO`, `KNI_KMOD`, `LIBRTE_KNI` and `LIBRTE_PMD_KNI` was removed in previous DPDK releases (after 19.11), so we don't need to do any modification for this
- Adding the proper argument in meson setup to include support for mlx5 driver
- Updating testpmd-run, `-w` is now `-a`, rest is the same
- Bumping img version to 0.2.8
- In cnf-app-mac-operator, the command used to retrieve MAC addresses from testpmd-app pods needed to be changed because the regex applied was not working, using the correct format now and adding more logs to better troubleshoot in the future

### CODE REVIEW

**Comment by ramperher** [comment]
This execution looks promising: https://www.distributed-ci.io/jobs/78e10d00-96bf-4237-ac43-bbdc2af7ce13/jobStates?sort=date

- all example-cnf resources were installed fine
- test execution went fine

but the test results throws a -300% of packet loss, which is a really bad figure...

```
2024-02-14 15:01:28,932 - run-trex - INFO -
Packets lost from 0 to 1: -3607768 packets, which is -300.64733333333334% packet loss
2024-02-14 15:01:28,932 - run-trex - INFO - Packets lost from 1 to 0: -3607767 packets, which is -300.64725% packet loss                                                                                                                     
2024-02-14 15:01:28,932 - run-trex - INFO - Total packets lost: -7215535 packets, which is -300.64729166666666% packet loss                                                                                                                  
2024-02-14 15:01:28,943 - run-trex - INFO - CR to be utilized for event creation: {'apiVersion': 'events.k8s.io/v1', 'kind': 'Event', 'metadata': {'name': 'trex-app-i8h4bn', 'namespace': 'example-cnf', 'ownerReferences': [{'apiVersion': 'examplecnf.openshift.io/v1', 'kind': 'TRexApp', 'name': 'trex-app', 'uid': 'f5e3e853-f0d4-4168-ab43-8bb8bd86fa2b', 'controller': True}]}, 'type': 'Normal', 'eventTime': '2024-02-14T15:01:28.908592Z', 'series': {'lastObservedTime': '2024-02-14T15:01:28.908592Z', 'count': 2}, 'reason': 'TestPassed', 'action': 'TestPassed', 'note': 'Test has passed with no packet loss, total packets: 2400000', 'regarding': {'namespace': 'example-cnf', 'kind': 'TRexApp', 'name': 'trex-app', 'uid': 'f5e3e853-f0d4-4168-ab43-8bb8bd86fa2b'}, 'reportingController': 'pod/job-trex-app-xbh9d', 'reportingInstance': 'trex-app'}                                                                                                             
2024-02-14 15:01:28,953 - run-trex - DEBUG -
Disconnecting from server at 'trex-server':'4501'           [SUCCESS]

Test has passed :-)
```

I'll move this change to in progress, so that people can review, but I'd like to hear your opinion on this, @capolrik @tkrishtop

**Comment by ramperher** [comment]
I've been comparing a job where we were using DPDK 19.11 with the last passed one with DPDK 23.11, and the only difference I can explicitly see is this -300% packet loss ratio we are having now, but the trex test is passing and testpmd-app pods seem to be forwarding the packages correctly. Let's see the comparison.

I've used these two jobs [(1)](https://www.distributed-ci.io/jobs/e880f4da-0e29-4194-94dd-1bd7e32db75e) and [(2)](https://www.distributed-ci.io/jobs/7fb91fdc-a523-4223-a0d1-ace8cc7ad59f), with DPDK 19.11 (from yesterday's daily jobs) to show you two different cases we are having. Case (2) is more or less similar to what it is happening with [DPDK 23.11](https://www.distributed-ci.io/jobs/78e10d00-96bf-4237-ac43-bbdc2af7ce13), but with a lower negative packet loss ratio.

Insights here:

- Taking a look to trexconfig-<x>.log file, in all cases, final port stats offer a similar order of magnitude for ipackets, ibytes and ierrors, nothing weird there. Example from (1)

```
-Per port stats table 
      ports |               0 |               1 
 -----------------------------------------------------------------------------------------
   opackets |               0 |               0 
     obytes |               0 |               0 
   ipackets |       356618820 |       356541799 
     ibytes |     25676555248 |     25671009600 
    ierrors |       325086177 |       328785900 
    oerrors |               0 |               0 
      Tx Bw |       0.00  bps |       0.00  bps 
```

- In trex-app.log, we can see the result of executing trex job at the end of the logs. It is calculating packet loss from 0-1 and from 1-0, and applying the average between them to calculate the total packet loss ratio. Normally, whereas one packet loss figure is usually positive, and normally with a figure of 6 orders of magnitude for the number of packets that are lost, the other one is usually negative and really similar in absolute value, so final packet loss is around 0%. Example from (1):

```
2024-02-14 18:51:42,934 - run-trex - INFO - 
Packets lost from 0 to 1: 195994 packets, which is 16.332833333333333% packet loss
2024-02-14 18:51:42,934 - run-trex - INFO - Packets lost from 1 to 0: -200858 packets, which is -16.738166666666668% packet loss
2024-02-14 18:51:42,935 - run-trex - INFO - Total packets lost: -4864 packets, which is -0.20266666666666666% packet loss
2024-02-14 18:51:42,945 - run-trex - INFO - CR to be utilized for event creation: {'apiVersion': 'events.k8s.io/v1', 'kind': 'Event', 'metadata': {'name': 'trex-app-pycm87', 'namespace': 'example-cnf', 'ownerReferences': [{'apiVersion': 'examplecnf.openshift.io/v1', 'kind': 'TRexApp', 'name': 'trex-app', 'uid': '368e2823-c66e-4f2b-ac18-5e4c98626906', 'controller': True}]}, 'type': 'Normal', 'eventTime': '2024-02-14T18:51:42.914392Z', 'series': {'lastObservedTime': '2024-02-14T18:51:42.914392Z', 'count': 2}, 'reason': 'TestPassed', 'action': 'TestPassed', 'note': 'Test has passed with no packet loss, total packets: 2400000', 'regarding': {'namespace': 'example-cnf', 'kind': 'TRexApp', 'name': 'trex-app', 'uid': '368e2823-c66e-4f2b-ac18-5e4c98626906'}, 'reportingController': 'pod/job-trex-app-qqhgd', 'reportingInstance': 'trex-app'}
2024-02-14 18:51:42,957 - run-trex - DEBUG - 
Disconnecting from server at 'trex-server':'4501'           [SUCCESS]
```

However, in (2), we had a case where the difference between both absolute values is higher, having a -27% packet loss:

```
2024-02-14 11:58:38,438 - run-trex - INFO - 
Packets lost from 0 to 1: -770283 packets, which is -64.19025% packet loss
2024-02-14 11:58:38,438 - run-trex - INFO - Packets lost from 1 to 0: 117438 packets, which is 9.7865% packet loss
2024-02-14 11:58:38,438 - run-trex - INFO - Total packets lost: -652845 packets, which is -27.201875% packet loss
2024-02-14 11:58:38,444 - run-trex - INFO - CR to be utilized for event creation: {'apiVersion': 'events.k8s.io/v1', 'kind': 'Event', 'metadata': {'name': 'trex-app-q0zcet', 'namespace': 'example-cnf', 'ownerReferences': [{'apiVersion': 'examplecnf.openshift.io/v1', 'kind': 'TRexApp', 'name': 'trex-app', 'uid': '243ecb5a-9487-4e21-b343-f262e6df35e4', 'controller': True}]}, 'type': 'Normal', 'eventTime': '2024-02-14T11:58:38.414071Z', 'series': {'lastObservedTime': '2024-02-14T11:58:38.414071Z', 'count': 2}, 'reason': 'TestPassed', 'action': 'TestPassed', 'note': 'Test has passed with no packet loss, total packets: 2400000', 'regarding': {'namespace': 'example-cnf', 'kind': 'TRexApp', 'name': 'trex-app', 'uid': '243ecb5a-9487-4e21-b343-f262e6df35e4'}, 'reportingController': 'pod/job-trex-app-z6src', 'reportingInstance': 'trex-app'}
2024-02-14 11:58:38,457 - run-trex - DEBUG - 
Disconnecting from server at 'trex-server':'4501'           [SUCCESS]
```

In DPDK 23.11, both figures were negative, and the number of lost packets had 7 orders of magnitude, that's a real difference:

```
2024-02-14 15:01:28,932 - run-trex - INFO - 
Packets lost from 0 to 1: -3607768 packets, which is -300.64733333333334% packet loss
2024-02-14 15:01:28,932 - run-trex - INFO - Packets lost from 1 to 0: -3607767 packets, which is -300.64725% packet loss
2024-02-14 15:01:28,932 - run-trex - INFO - Total packets lost: -7215535 packets, which is -300.64729166666666% packet loss
2024-02-14 15:01:28,943 - run-trex - INFO - CR to be utilized for event creation: {'apiVersion': 'events.k8s.io/v1', 'kind': 'Event', 'metadata': {'name': 'trex-app-i8h4bn', 'namespace': 'example-cnf', 'ownerReferences': [{'apiVersion': 'examplecnf.openshift.io/v1', 'kind': 'TRexApp', 'name': 'trex-app', 'uid': 'f5e3e853-f0d4-4168-ab43-8bb8bd86fa2b', 'controller': True}]}, 'type': 'Normal', 'eventTime': '2024-02-14T15:01:28.908592Z', 'series': {'lastObservedTime': '2024-02-14T15:01:28.908592Z', 'count': 2}, 'reason': 'TestPassed', 'action': 'TestPassed', 'note': 'Test has passed with no packet loss, total packets: 2400000', 'regarding': {'namespace': 'example-cnf', 'kind': 'TRexApp', 'name': 'trex-app', 'uid': 'f5e3e853-f0d4-4168-ab43-8bb8bd86fa2b'}, 'reportingController': 'pod/job-trex-app-xbh9d', 'reportingInstance': 'trex-app'}
2024-02-14 15:01:28,953 - run-trex - DEBUG - 
Disconnecting from server at 'trex-server':'4501'           [SUCCESS]
```

- There's not a big difference in terms of results in testpmd-app logs (which is, in fact, the pod we're changing in this change, but we observe less RX-missed packages in DPDK 23.11. Here we have the case of (1), at the end of the execution:

```
Port statistics ====================================
  ######################## NIC statistics for port 0  ########################
  RX-packets: 6588144866 RX-missed: 1258254    RX-bytes:  447993850612
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 6586883557 TX-errors: 0          TX-bytes:  421560547520

  Throughput (since last show)
  Rx-pps:      9624032          Rx-bps:   5235476800
  Tx-pps:      9624022          Tx-bps:   4927502928
  ############################################################################

  ######################## NIC statistics for port 1  ########################
  RX-packets: 6588400690 RX-missed: 1037338    RX-bytes:  448011246576
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 6587361427 TX-errors: 0          TX-bytes:  421591131264

  Throughput (since last show)
  Rx-pps:      9621585          Rx-bps:   5234142936
  Tx-pps:      9621627          Tx-bps:   4926274016
  ############################################################################
```

Whereas in DPDK 23.11, we had 1 order of magnitude less in RX/TX pkts, but RX-missed is considerably lower (3 orders of magnitude lower).

```
  ######################## NIC statistics for port 0  ########################
  RX-packets: 228610908  RX-missed: 1254       RX-bytes:  14631183160
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 228610908  TX-errors: 0          TX-bytes:  14631098112

  Throughput (since last show)
  Rx-pps:            0          Rx-bps:            0
  Tx-pps:            0          Tx-bps:            0
  ############################################################################

  ######################## NIC statistics for port 1  ########################
  RX-packets: 228610908  RX-missed: 1254       RX-bytes:  14631183160
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 228610908  TX-errors: 0          TX-bytes:  14631098112

  Throughput (since last show)
  Rx-pps:            0          Rx-bps:            0
  Tx-pps:            0          Tx-bps:            0
  ############################################################################
```

- In loadbalancer logs, we can see differences between the cases where packet loss is around 0% and higher (in negative). E.g. in case (1), where packet loss was nearly 0%, what it is tx in ports 0-1 is rx in ports 2-3:

```
Port statistics ====================================
  ######################## NIC statistics for port 0  ########################
  RX-packets: 498843781  RX-missed: 5925       RX-bytes:  33921377020
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 50         TX-errors: 0          TX-bytes:  3200

  Throughput (since last show)
  Rx-pps:            0          Rx-bps:            0
  Tx-pps:            0          Tx-bps:            0
  ############################################################################

  ######################## NIC statistics for port 1  ########################
  RX-packets: 498826941  RX-missed: 5826       RX-bytes:  33920231900
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 50         TX-errors: 0          TX-bytes:  3200

  Throughput (since last show)
  Rx-pps:            0          Rx-bps:            0
  Tx-pps:            0          Tx-bps:            0
  ############################################################################

  ######################## NIC statistics for port 2  ########################
  RX-packets: 50         RX-missed: 0          RX-bytes:  3200
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 498839339  TX-errors: 0          TX-bytes:  31925717696

  Throughput (since last show)
  Rx-pps:            0          Rx-bps:            0
  Tx-pps:         1304          Tx-bps:       667664
  ############################################################################

  ######################## NIC statistics for port 3  ########################
  RX-packets: 50         RX-missed: 0          RX-bytes:  3200
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 498822534  TX-errors: 0          TX-bytes:  31924642176

  Throughput (since last show)
  Rx-pps:            0          Rx-bps:            0
  Tx-pps:         1248          Tx-bps:       639216
  ############################################################################
```

However, this is not true in the other cases, with a higher, negative, packet loss. E.g. from (2), rx pkts in ports 2-3 are considerably higher (DPDK 23.11 follows the same trend):

```
Port statistics ====================================
  ######################## NIC statistics for port 0  ########################
  RX-packets: 5069032450 RX-missed: 380547     RX-bytes:  344694206448
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 78         TX-errors: 0          TX-bytes:  4992

  Throughput (since last show)
  Rx-pps:      4127074          Rx-bps:   2245129920
  Tx-pps:            0          Tx-bps:            0
  ############################################################################

  ######################## NIC statistics for port 1  ########################
  RX-packets: 5069418904 RX-missed: 380423     RX-bytes:  344720485120
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 78         TX-errors: 0          TX-bytes:  4992

  Throughput (since last show)
  Rx-pps:      4126484          Rx-bps:   2244807776
  Tx-pps:            0          Tx-bps:            0
  ############################################################################

  ######################## NIC statistics for port 2  ########################
  RX-packets: 333859     RX-missed: 332766     RX-bytes:  22702100
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 5048943839 TX-errors: 0          TX-bytes:  323132405632

  Throughput (since last show)
  Rx-pps:            0          Rx-bps:            0
  Tx-pps:      4122155          Tx-bps:   2110543104
  ############################################################################

  ######################## NIC statistics for port 3  ########################
  RX-packets: 333027     RX-missed: 331934     RX-bytes:  22645524
  RX-errors: 0
  RX-nombuf:  0         
  TX-packets: 5049204270 TX-errors: 0          TX-bytes:  323149072320

  Throughput (since last show)
  Rx-pps:            0          Rx-bps:            0
  Tx-pps:      4123145          Tx-bps:   2111048944
  ############################################################################
```

So, my conclusions from this analysis (not being an expert on trex data analysis) are that:

- testpmd-app is working, there's MAC forwarding being applied
- packet loss is still negative, but higher in absolute value
- the trends in loadbalancer statistics are similar than cases with a packet loss around -20%

To conclude; unless anybody has more experience and knowledge about how to understand these results, I consider DPDK 23.11 is behaving well, and maybe it's worth to test this on a daily basis to see if this is repeated or just an isolated case (anyway I'm testing again to see if the same happens).

**Comment by ramperher** [comment]
We commented on the daily to test without the loadbalancer, to see how it goes.

I ran a new test without setting up load balancer (this means `enable_lb: false`), to see if MAC forwarding mode is working fine, but trex pods were not created properly, so the job will fail. I've seen these issues:

```
$ oc logs job-trex-app-4zcw9
...
+ (( count-- ))
+ [[ 1 -gt 0 ]]
+ nc -z trex-server 4500
+ sleep 2
+ (( count-- ))
+ [[ 0 -gt 0 ]]
+ [[ 0 -gt 0 ]]
+ echo_ts 'ERROR: trex-server could not start properly, exiting... '
++ date -u +%Y-%m-%dT%H:%M:%S%z
2024-02-15T16:13:25+0000 ERROR: trex-server could not start properly, exiting...
+ echo '2024-02-15T16:13:25+0000 ERROR: trex-server could not start properly, exiting... '
+ exit 1

$ oc logs trexconfig-75fdcfc845-tk8m6
configure webserver
try to start webserver in port:  8096
INPUT: core_string: 1-3,41-43
pcis = 0000:37:02.1,0000:37:02.3
{'name': 'intel-numa0-net1', 'mac': '20:04:0f:f1:89:01', 'namespace': 'example-cnf'}
{'name': 'intel-numa0-net1', 'mac': '20:04:0f:f1:89:02', 'namespace': 'example-cnf'}
worker-0
destination mach fetched
[{'version': 2, 'interfaces': ['37:02.1', '37:02.3'], 'port_info': [{'dest_mac': 'link state change event', 'src_mac': '20:04:0f:f1:89:01'}, {'dest_mac': 'link state change event', 'src_mac': '20:04:0f:f1:89:02'}], 'c': 4, 'platform': {'master_thread_id': 1, 'latency_thread_id': 2, 'dual_if': [{'socket': 0, 'threads': [3, 41, 42, 43]}]}}]
2024-02-15T16:14:15+0000 Starting TRex server ...
+ CORE_COUNT=4
+ '[' -z 4 ']'
+ cd /opt/trex/trex-core/scripts
+ n=0
+ '[' 0 -ge 5 ']'
+ ./_t-rex-64 -i --no-hw-flow-stat -c 4
Starting  TRex v2.85 please wait  ...
String of dest MAC should be in format '12:34:56:78:9a:bc'.
+ n=1
+ sleep 5
+ '[' 1 -ge 5 ']'
+ ./_t-rex-64 -i --no-hw-flow-stat -c 4
Starting  TRex v2.85 please wait  ...
String of dest MAC should be in format '12:34:56:78:9a:bc'.
+ n=2
+ sleep 5
+ '[' 2 -ge 5 ']'
+ ./_t-rex-64 -i --no-hw-flow-stat -c 4
Starting  TRex v2.85 please wait  ...
String of dest MAC should be in format '12:34:56:78:9a:bc'.
+ n=3
+ sleep 5
+ '[' 3 -ge 5 ']'
+ ./_t-rex-64 -i --no-hw-flow-stat -c 4
Starting  TRex v2.85 please wait  ...
String of dest MAC should be in format '12:34:56:78:9a:bc'.
+ n=4
+ sleep 5
+ '[' 4 -ge 5 ']'
+ ./_t-rex-64 -i --no-hw-flow-stat -c 4
Starting  TRex v2.85 please wait  ...
String of dest MAC should be in format '12:34:56:78:9a:bc'.
+ n=5
+ sleep 5
+ '[' 5 -ge 5 ']'
+ echo_ts 'Failed to start TRex Server on 5 retries...'
++ date -u +%Y-%m-%dT%H:%M:%S%z
2024-02-15T16:14:40+0000 Failed to start TRex Server on 5 retries...
+ echo '2024-02-15T16:14:40+0000 Failed to start TRex Server on 5 retries...'
```

I need to recheck this, to see what's going on.

**Comment by manurodriguez** [comment]
> In the last job, it happened just the same effect, again a packet loss ratio of -300%
> 
> I know we had pending to really understand what a negative packet loss ratio means, but having these results, I don't know if it's better to consider this change to be merged and then continue analyzing in parallel the trex code we are using to calculate the results, or the opposite.
>

I do not know much about example-cnf but from my side, if having negative packet loss won't make the job fail then I'm ok with moving forward, and once we have all components using DPDK 23.11 then we can revisit.

**Review by manurodriguez — APPROVED** [review]
:+1:

**Comment by ramperher** [comment]
The direct mode is working in the current example-cnf code: https://www.distributed-ci.io/jobs/27f98cf7-c67a-4a86-9db5-bd3c7b78cc60, but not working with DPDK 23.11: https://www.distributed-ci.io/jobs/20e767de-ad2b-4ea9-bdfa-b4dcf5c902f0/jobStates

I'd prefer to wait till fixing this issue for this version, because if direct mode is not working, then it means there's probably something wrongly configured.

I'll move this to draft status

**Comment by ramperher** [comment]
I finally managed to find the issue that we were having when testing direct mode with this change. In fact, here we have a test using the direct mode with the current DPDK 19.11, and it passed: https://www.distributed-ci.io/jobs/27f98cf7-c67a-4a86-9db5-bd3c7b78cc60/jobStates?sort=date, and if you check trex-app.log file, packet loss is 0%

The problem was the parsing of the MAC addresses of the testpmd-app pods from cnf-app-mac-operator, which is currently performed with this command: https://github.com/openshift-kni/example-cnf/blob/main/cnf-app-mac-operator/internal/controller/cnfappmac_controller.go#L433

However, if you check in any failed job with DPDK 23.11 using direct mode (e.g. https://www.distributed-ci.io/jobs/3501887b-3ae3-4211-85c3-331804d9b1d5), you can see there are more file lines that match that regex, and are related to this "link state change" that is eventually captured by the cnf-app-mac-operator code.

I've changed that code to properly capture the MAC address, and fail if not found any MAC address.

I launched a job and it passed, and packet loss is 0% - https://www.distributed-ci.io/jobs/5f2f2f56-45e2-466b-97d2-54b6b0ac2676/jobStates?sort=date (this is about to finish and be reported here). So this demonstrates that DPDK 23.11 is working fine and it's not the root cause of the -300% packet loss ratio we have when using the loadbalancer.

So, having said that, I believe it's fine to merge this, wdyt? @capolrik @tkrishtop @manurodriguez , also adding @tonyskapunk who commented to run more tests before merging this.

File: cnf-app-mac-operator/internal/controller/cnfappmac_controller.go
Comment (manurodriguez) [inline cnf-app-mac-operator/internal/controller/cnfappmac_controller.go:449]: Nice catch, I was reading that this regex could also match octets separated either by colon or dash, for example: 
```
manuel@fedora:~$ echo "3D-F2-C9:A6-B3:4F" | egrep '([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})$'
3D-F2-C9:A6-B3:4F
```
So maybe we can be a little bit more strict and allow only MACs with octets separated by colons only:
```
manuel@fedora:~$ echo "3D-F2-C9:A6-B3:4F" | egrep '([0-9A-Fa-f]{2}[:]){5}([0-9A-Fa-f]{2})$'
manuel@fedora:~$ echo "3D:F2:C9:A6:B3:4F" | egrep '([0-9A-Fa-f]{2}[:]){5}([0-9A-Fa-f]{2})$'
3D:F2:C9:A6:B3:4F
```
Other than that, looks good to me

**Review by manurodriguez — APPROVED** [review]

File: cnf-app-mac-operator/internal/controller/cnfappmac_controller.go
Comment (ramperher) [reply cnf-app-mac-operator/internal/controller/cnfappmac_controller.go:449]: sure, let's do that

**Comment by ramperher** [comment]
If there are no objections to merge this change, I'll do it tomorrow. I'll be waiting for more feedback in the meantime. Thanks!

**Review by capolrik — APPROVED** [review]
Nice work!

**Pull request merged**
